apps/rig_kafka/lib/rig_kafka/client.ex

```diff
@@ -28,11 +30,82 @@ defmodule RigKafka.Client do
       {:ok, state}
     end
 
+    # TODO: move to separate file
+
+    def extract_headers(%{
+          "cloudEvents_contentType" => contentType,
+          "cloudEvents_cloudEventsVersion" => cloudEventsVersion,
+          "cloudEvents_source" => source,
+          "cloudEvents_eventType" => eventType,
+          "cloudEvents_eventTime" => eventTime,
+          "cloudEvents_eventID" => eventID
+        }) do
+      %{
+        "contentType" => contentType,
+        "cloudEventsVersion" => cloudEventsVersion,
+        "source" => source,
+        "eventType" => eventType,
+        "eventTime" => eventTime,
+        "eventID" => eventID
+      }
+    end
+
+    def extract_headers(%{
+          "cloudEvents_contenttype" => contentType,
+          "cloudEvents_specversion" => cloudEventsVersion,
+          "cloudEvents_source" => source,
+          "cloudEvents_type" => eventType,
+          "cloudEvents_time" => eventTime,
+          "cloudEvents_id" => eventID
+        }) do
+      %{
+        "contenttype" => contentType,
+        "specversion" => cloudEventsVersion,
+        "source" => source,
+        "type" => eventType,
+        "time" => eventTime,
+        "id" => eventID
+      }
+    end
+
+    def extract_headers(_headers), do: %{}
+
     @impl :brod_group_subscriber
     def handle_message(topic, partition, message, %{callback: callback} = state) do
-      %{offset: offset, value: body} = Enum.into(kafka_message(message), %{})
+      %{offset: offset, value: body, headers: headers} = Enum.into(kafka_message(message), %{})
+
+      deconstructed_headers =
+        headers
+        |> Enum.into(%{})
+        |> extract_headers()
+
+      decoded_body =
+        cond do
+          Map.get(deconstructed_headers, "contentType") == "avro/binary" ->
+            data = Jason.decode!(Serializer.decode_body(body, "avro"))
```

Comment:
- if `Serializer.decode_body` fails, wouldn't it return an ok or error tuple?
- would it be reasonable to fall-back to another format in case decoding fails?

Reply:
Hmm, good point, I'll take a look.